vm/devices/storage/disk_get_vmgs/src/lib.rs

```diff
@@ -224,6 +224,19 @@ impl DiskIo for GetVmgsDisk {
             .await
             .map_err(|err| DiskError::Io(io::Error::new(io::ErrorKind::Other, err)))
     }
+
+    async fn unmap(
+        &self,
+        _sector: u64,
+        _count: u64,
+        _block_level_only: bool,
+    ) -> Result<(), DiskError> {
+        Ok(())
+    }
+
+    fn unmap_behavior(&self) -> disk_backend::UnmapBehavior {
+        disk_backend::UnmapBehavior::Ignored
```

Comment:
fwiw, the nice thing about the IDET is that it gives you a compile time way to encode the notion of "if Unmap is Ignored, don't call the `unmap` method". i.e: these two methods can instead be:

```rust
async fn supports_unmap(&mut self) -> disk_backend::UnmapBehavior<'_>;

// where
enum UnmapBehavior<'a> {
    Ignored,
    Unspecified(&'a dyn Unmap),
    Zeroes(&'a dyn Unmap),
}
```

this ensures callers of unmap must consider the various unmap behaviors

Reply:
or, in other words: IDETs aren't just about making certain extensions optional, but rather, a more generic mechanism of expressing bounds on when certain APIs can be called

Reply:
I'll note that the current formulation is quite similar to what I consider to be the `fn method() { unimplemented!() }` + `fn supports_method() -> bool` anti-pattern, where the compiler doesn't help catch callers who neglect to check for method support prior to calling the method.

not to say there aren't times when you want to do that (especially if you care about perf, and the additional deref is something you want to avoid), but idk if that's the case with the Unmap case

Reply:
I don't think there's value on the caller side (`Disk`) in this case. The caller can't tell the difference between ignored and unspecified, so there's not really any value in having the caller be forced to go through this dance _at the time it's calling unmap_--by that point, it's too late to do something different. And so it's just more convenient to always let the caller call unmap, even if it's ignored.

The caller may want to know _ahead of time_ what the unmap behavior is, e.g., because it wants to tell the guest that there's no point in sending unmap SCSI commands, and the IDET pattern doesn't improve the ergonomics of determining that. It's handy to be able to get a behavior enum value that you can store (has no lifetime).

On the implementation side (`DiskIo`), it's true that IDET would give us a mild improvement in enforcing that ignored is really ignored. But I think the code for things like disk_layered and disk_striped end up being more annoying to write.

And another oddity is that, because with the IDET pattern you still end up implementing the trait for your (pub) type, you still have to consider what happens if someone calls unmap directly, even if you return `None`/`Ignored` from your IDET method.

So there I just don't think it's worth it.

Reply:
Yeah, the fact that the "driver" in charge of calling this API isn't the caller directly, but rather, something like SCSI commands from the Guest is a good point.

> And another oddity is that, because with the IDET pattern you still end up implementing the trait for your (pub) type, you still have to consider what happens if someone calls unmap directly, even if you return `None`/`Ignored` from your IDET method.

This is a fair point. This doesn't really happen in the Option case, given that you simply wouldn't implement the trait... but with Ignored things are a bit weird, for sure.